File: 2018/day02/src/main.jl
function first_part(input)
    twice, thrice = 0, 0

    for line in String.(split(input))
        freqs = Array{Int8}(undef, 256)

        for i in line
            freqs[Int8(i)] += 1
        end

        if any(x -> x == 2, freqs)
            twice += 1
        end
        if any(x -> x == 3, freqs)
            thrice += 1
        end
    end

    println(twice * thrice)
end

function second_part(args)
    ids = collect(String.(split(input)))

    for i in ids, j in ids
        # println(i," ", j)
        common = find_common_letters(i, j)
        if !isempty(common)
            println(common)
            break
        end
    end

end

function find_common_letters(id1, id2)
    if id1 == id2 || length(id1) != length(id2)
        return []
    end

    found_once = false
    ids_tuple = zip(id1, id2)
    for (c1, c2) in ids_tuple
        if c1 != c2
            if found_once
                return []
            end
            found_once = true
        end
    end

    return collect(
        map((c, _) -> c,
        filter((c1, c2) -> c1 == c2,
        ids_tuple)))
end


file = open("../input/input.txt")
input = read(file, String)

first_part(input)
second_part(input)

close(file)

File: 2018/day02/src/main.rs
use std::{
    error::Error,
    io::{ Read, stdin }
};

fn first_part(input: &str) -> Result<(), Box<Error>> {
    let (mut twice, mut thrice) = (0, 0);

    for line in input.lines() {
        let mut freqs = [0u8; 256];

        for b in line.as_bytes().iter().map(|&x| x as usize) {
            freqs[b] += 1;
        }

        if freqs.iter().any(|&x| x == 2) {
            twice += 1;
        }
        if freqs.iter().any(|&x| x == 3) {
            thrice += 1;
        }
    }
    println!("{}", twice * thrice);

    Ok(())
}

fn second_part(input: &str) -> Result<(), Box<Error>> {
    let ids: Vec<&str> = input.lines().collect();

    for i in ids.iter() {
        for j in ids.iter() {
            if let Some(common) = find_common_letters(&i, &j) {
                println!("{}", common);
                return Ok(());
            }
        }
    }

    Err(From::from("Didn't found the boxes"))
}

fn find_common_letters(id1: &str, id2: &str) -> Option<String> {
    if id1 == id2 || id1.len() != id2.len() {
        return None;
    }

    let mut found_once = false;
    for (c1, c2) in id1.chars().zip(id2.chars()) {
        if c1 != c2 {
            if found_once {
                return None;
            }

            found_once = true;
        }
    }

    Some(
        id1.chars().zip(id2.chars())
            .filter(|&(c1, c2)| c1 == c2)
            .map(|(c, _)| c)
            .collect()
    )
}

fn main() -> Result<(), Box<Error>> {
    let mut input = String::new();
    stdin().read_to_string(&mut input)?;

    first_part(&input)?;
    second_part(&input)?;

    Ok(())
}
